Blog Post: Announcing  the OTel Arrow Protocol (#3436)
open-telemetry/opentelemetry.io: content/en/blog/2023/otel-arrow/index.md

release of a new pair of receiver/exporter in the OTel contrib repository that
supports it. This protocol, designed to complement the OTLP protocol in
situations with substantial telemetry data volumes, has been under discussion
and development for two years. It represents a collaborative effort between
[F5](https://f5.com),
[ServiceNow Cloud Observability](https://www.servicenow.com/products/observability.html)
, and numerous OTel technical leaders (see
[donation](https://github.com/open-telemetry/community/issues/1332)). The
compression benefits are significant, demonstrating a **2x improvement** for the
majority of workloads and even greater enhancements for workloads that contain
multivariate metrics, which share attributes. One of the highlights of the OTel
Arrow integration is its seamless adaptability. In a typical deployment,
incorporating OTel Arrow doesn't necessitate any substantial changes. Users
simply need to redeploy a new version of the collector with a slightly tweaked
configuration.

This new protocol will serve as the cornerstone for future advancements in
telemetry data processing and will generally foster enhanced integration with

> modern telemetry backends.

Author: Good point. I will change the intro as follow:

We are thrilled to unveil the **OTel Arrow Protocol**, which is based on [Apache Arrow](https://arrow.apache.org/), a columnar-oriented memory format used for developing analytics applications. We are also excited to announce the release of a new pair of receiver/exporter in the [opentelemetry-collector-contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib/) repository that supports this protocol.

batch of metrics with shared attributes without redundantly replicating these
attributes. In specific scenarios, this redundancy places an undue strain on
both the network bandwidth and overall resource utilization. Our newly designed
protocol addresses this by offering an enhanced representation of multivariate
metrics. In certain scenarios, we've seen compression improvements of up to 7x
compared to OTLP, all without modifications on the client side. Future client
SDKs could be implemented to seamlessly expose this enhancement, potentially
leading to even better results for applications.

The advancements don't end there. In a subsequent phase of this project, we aim
to leverage the columnar layout to significantly enhance data processing speeds
within an expanded OTel Collector architecture that natively supports a new
Arrow-based pipeline. Based on our proof of concept, we anticipate at least an
order of magnitude improvement in data processing speed with this updated
collector.

These different sources of inefficiency and misalignment are the rationale
behind our support for the new OTel Arrow Protocol as an alternative to the

> existing OTLP protocol. Our decision to leverage the well known Apache Arrow

Author: Done

spaces. Its rich ecosystem boasts a range of powerful libraries and tools, from
Parquet bridges to query engines such as DataFusion. Such resources can expedite
the introduction of innovative features, aligning OpenTelemetry more closely
with modern data pipelines that are increasingly pivoting towards Apache Arrow.

A specification for this protocol (OTEP 0156) can be found
[here](https://github.com/open-telemetry/oteps/blob/main/text/0156-columnar-encoding.md).
A reference implementation of the encoding/decoding function can be accessed
[here](https://github.com/open-telemetry/otel-arrow). Additionally, the new pair
of OTel receiver/exporter that supports this protocol is available in the
[contrib repository](https://github.com/open-telemetry/opentelemetry-collector-contrib).

## How can I leverage OTel Arrow in my deployment?

In the initial phase of this project, our primary goal is to optimize
communication between two collectors. This is commonly observed in setups where
telemetry traffic is funneled through one or multiple collectors before being
relayed across the internet for backend processing. Given the increased

> complexity of the OTel Arrow Protocol compared to the original OTLP Protocol,

Reviewer: ```suggestion
complexity of the OTel Arrow Protocol compared to the original OTLP,
```

its containment between two collectors offers an easier target to hit and
reduces potential disruptions to the broader ecosystem. Existing client SDKs,
processors, receivers and exporters can continue to work seamlessly. Only the
exporter and the receiver between the two collectors need to be reconfigured.
The immediate benefit will be a reduction in network bandwidth, leading to
direct savings on network costs (up to 7x for metrics, 2x for logs, and traces).
For a comprehensive breakdown of this deployment, click here [add link].

![Internet Traffic Reduction](./traffic-reduction.png)

As is often the case, there isn't a one-size-fits-all solution. Deployments with
limited resources or those generating minimal telemetry should stick with the
standard collector based on OTLP. Moreover, OTel Arrow necessitates the support
of bi-directional gRPC streams and some degree of batching to fully benefit from
the columnar representation. This can make the solution unsuitable in certain
specific scenarios. It's also worth noting that, during this phase of the
project, a slight increase in CPU and memory usage is expected in the collector.

> This is due to the overhead of automatically translating OTLP objects to OTel
> Arrow objects. However, this overhead will be completely eliminated in the

Reviewer: ```suggestion
This is due to the overhead of automatically translating OTLP objects to OTel
Arrow Protocol objects. However, this overhead will be completely eliminated in the
```

Your deployments and feedback will assist us in further strengthening this
project.

## Next

In the future, we plan to focus on fully integrating OTel Arrow throughout the
ecosystem. The proposed developments are as follows (not necessarily in order of
priority):

- **Development of Client SDKs with native support for OTel Arrow and
  multivariate metrics**: Aimed at optimizing scenarios with high telemetry
  producers.
- **Introduction of a new pipeline type in the Collector**: This will introduce
  a new generation of receiver, processor, and exporter, each designed
  specifically for consuming and/or producing OTel Arrow messages. By
  streamlining communication between components, we expect to enhance data
  processing efficiency. Significant acceleration is anticipated as conversions
  to and from OTLP, as well as the serialization and deserialization of
  telemetry batches, will be bypassed.
- **Leveraging SIMD-based data processing engines from the Apache Arrow
  ecosystem**: This will further accelerate telemetry data processing and expand
  the range of data processing capabilities.
- **Consideration of adding Parquet exporters**: Made possible by the existing

>   bridge between Arrow and Parquet.

Author: Done

---

> title: Reduce Telemetry Traffic by Up to 2 Times Using OTel Arrow Protocol

Reviewer 2: "Reduce" and "up to 2 times" doesn't flow well. How about a shorter title like:

```suggestion
title: Reduce Telemetry Traffic by Half Using the OTel Arrow Protocol
```

Author: @jmacd and I ended up choosing a different title.

linkTitle: OTel Arrow Protocol

> date: 2023-09-14

Reviewer 2: Let's bring the date a little closer to now :)

```suggestion
date: 2023-11-01
```

author: >-
  [Laurent Quérel](https://github.com/lquerel) (F5), [Joshua
  MacDonald](https://github.com/jmacd) (ServiceNow)

> ---

Reviewer 2: This should fix the cspell warning.

```suggestion
cSpell:ignore: Quérel
---
```

been under discussion and development for two years. It represents a
collaborative effort between [F5](https://f5.com),
[ServiceNow Cloud Observability](https://www.servicenow.com/products/observability.html)
, and numerous technical leaders from the OpenTelemetry community (see
[donation](https://github.com/open-telemetry/community/issues/1332)). The
compression benefits are significant, demonstrating a **2x improvement
compression efficiency** for the majority of workloads and even greater
enhancements for workloads that contain multivariate metrics, which share
attributes. One of the highlights of the OTel Arrow Protocol integration is its
seamless adaptability. In a typical deployment, incorporating OTel Arrow
Protocol doesn't necessitate any substantial changes. Users simply need to
redeploy a new version of the OpenTelemetry Collector with a slightly tweaked
configuration.

This new protocol will serve as the cornerstone for future advancements in
telemetry data processing and will generally foster enhanced integration with
modern telemetry backends.

## Why a new protocol?

The growth in telemetry data is undeniable and rapid. This surge is attributed
to several factors:

> - the proliferation of devices and sensors,
> - the shift from monolithic application deployment to more granular forms like
>   containers and serverless functions,
> - and the increasing reliance on data-driven and AI-driven technologies.

Reviewer 2: ```suggestion
- Proliferation of devices and sensors
- Shift from monolithic application deployment to more granular forms like
  containers and serverless functions
- Increasing reliance on data-driven and AI-driven technologies
```

negligible. In other cases, this initial overhead is very quickly offset after
just the first few batches. The columnar organization also lends itself better
to compression. For very large batch sizes, ZSTD does an excellent job as long
as the compression window is sufficiently large, but even in this case, the new
protocol remains superior. As previously mentioned, these compression gains can
be higher for traffic predominantly containing multivariate metrics.

![Avg % of Compressed Size Improvement of OpenTelemetry Protocol with Apache Arrow over OTLP](average-improvement-heatmap.png)

The advancements don't end there. In a subsequent phase of this project, we aim
to leverage the columnar layout to significantly enhance data processing speeds
within an expanded OpenTelemetry Collector architecture that natively supports a
new Arrow-based pipeline. Based on our proof of concept, we anticipate at least
an order of magnitude improvement in data processing speed with this updated
collector.

These different sources of inefficiency and misalignment are the rationale
behind our support for the new OpenTelemetry Protocol with Apache Arrow as an alternative to the

> existing OTLP protocol. Our decision to leverage the well known Apache Arrow

Reviewer: ```suggestion
existing OpenTelemetry Protocol. Our decision to leverage the well known Apache Arrow
```

OTLP Protocol is doubling Protocol

environments. This was done to assess both the collector's behavior and the
protocol's resilience when confronted with real traffic. Not only did these
deployments result in enhancements to our automated validation framework, but
they also substantiated our benchmark findings.

While we've diligently worked to identify and address regressions and issues, we
recognize the complex and varied nature of real-world scenarios. Thus, we
encourage the community to evaluate this new protocol across diverse deployment
situations and traffic loads, starting with the least sensitive environments.
Your deployments and feedback will assist us in further strengthening this
project.

To facilitate this testing phase, we have also improved the file exporter to
save and anonymize the data in a compressed format. We have developed tools to
enable the analysis and replay of this anonymized traffic in order to identify
either bugs or sources of inefficiency. This approach recently played an
important role in improving the logic that defines the best Arrow schemas to use
for a given traffic. During an experiment in a production environment, we

> noticed a degradation in the compression rate after 1 million spans. Upon

Reviewer: ```suggestion
noticed a degradation in the compression rate after one million spans. Upon
```
